### PR TITLE
Replacing the static references of SECP256K1 by an interface

### PR DESCRIPTION
**Pull request by daniel-iobuilders**
Signed-off-by: Daniel Lehrner <daniel@io.builders>

## PR description
At the moment Besu only supports the Ethereum standard SECP256K1 key algorithm. Due to NIST compliance, we would like to add support for SECP256R1 keys.
In this first PR all the static reference calls to SECP256K1 are replaced by an interface. This is a first step to allow different elliptic curves to be added by future PRs. 

## Fixed Issue(s)
-

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by atoulme** [comment]
This is an intriguing proposal, do you want to change the key wholesale across the whole system? Would you want to allow some compatibility for example at the network layer?

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/account/Account.java [hunk 1]
```diff
@@ -14,9 +14,7 @@
  */
 package org.hyperledger.besu.tests.acceptance.dsl.account;
 
-import org.hyperledger.besu.crypto.SECP256K1.KeyPair;
-import org.hyperledger.besu.crypto.SECP256K1.PrivateKey;
-import org.hyperledger.besu.crypto.SECP256K1.PublicKey;
+import org.hyperledger.besu.crypto.*;
```
Comment (shemnon) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/account/Account.java:17]: Spotless is configured to reject star imports.  Please replace with explicit imports.

File: crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignature.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.bouncycastle.math.ec.ECPoint;
+
+import java.math.BigInteger;
+import java.util.Optional;
+import java.util.function.UnaryOperator;
+
+public interface EllipticCurveSignature {
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignature.java:11]: I feel we should go a bit further than simply pulling out the interface methods.  I think we should reconsider all the APIs being pulled through and look to see if better homes exist (calculateECDHKeyAgreement, getHalfCureOrder, enableNarive, etc).

A higher level goal may be to provide SPI classes to make it work with javax.crypto and friends, rather than a single use interface.

I also think we can do better than the `EllipticCurveSignature` name.  Account sigs may not always be ECDSA (DSA and RSA are also listed in FIPS). So perhaps `SignatureAlgorithm`?  Or make it role based `AccountSignatures`.  Also the noun implies it is the signature itself, not the class that does the signing, verification, and key management.

File: crypto/src/main/java/org/hyperledger/besu/crypto/KeyPair.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.math.BigInteger;
+import java.security.KeyPairGenerator;
+import java.util.Arrays;
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class KeyPair {
+
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/KeyPair.java:15]: Should we just use `javax.security.KeyPair` and move the static method elsewhere?

File: crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignatureFactory.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.hyperledger.besu.crypto;
+
+public class EllipticCurveSignatureFactory {
+
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignatureFactory.java:4]: I'm presuming this is a placeholder for future work to add `secp256r1`. How will the choice of curve be determined?

File: crypto/src/main/java/org/hyperledger/besu/crypto/KeyPairUtil.java [hunk 5]
```diff
@@ -32,6 +32,7 @@
 
 public class KeyPairUtil {
   private static final Logger LOG = LogManager.getLogger();
+  private static final EllipticCurveSignature ELLIPTIC_CURVE_SIGNATURE = EllipticCurveSignatureFactory.getInstance();
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/KeyPairUtil.java:35]: Fixing this as a static final field makes runtime or startup determination of the curve problematic.  The config would be locked in at _class loading_ time.  We should use the method on the factory instead.  Locking the factory is one thing, locking what the factory produces is something entirely different.

If we had a lazy initialized value that would also allow for later than classloading configuration. `com.google.common.base.Suppliers#memoize` is a quick path to victory here.

File: crypto/src/main/java/org/hyperledger/besu/crypto/NodeKey.java [hunk 6]
```diff
@@ -22,25 +22,26 @@
 public class NodeKey {
 
   private final SecurityModule securityModule;
+  private final EllipticCurveSignature ellipticCurveSignature = EllipticCurveSignatureFactory.getInstance();
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/NodeKey.java:25]: See my notes in `KeyPairUtil`

File: crypto/src/main/java/org/hyperledger/besu/crypto/PrivateKey.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.units.bigints.UInt256;
+
+import java.math.BigInteger;
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class PrivateKey implements java.security.PrivateKey {
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/PrivateKey.java:11]: Name should be more descriptive, like `SECPPrivateKey.`

File: crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+import org.bouncycastle.math.ec.FixedPointCombMultiplier;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class PublicKey implements java.security.PublicKey {
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java:16]: Name should be more descriptive, like `SECPPublicKey`

**Comment by lucassaldanha** [comment]
> This is an intriguing proposal, do you want to change the key wholesale across the whole system?

The plan is to support different key algorithms across the board. We aren't aiming for interop between nodes using different key algorithms.

> Would you want to allow some compatibility for example at the network layer?

I don't think I follow. Would you mind elaborating on this?

File: crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    public static final int BYTE_LENGTH = 64;
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java:18]: We shouldn't presume a key length.  What if secp256r1 is un-recommened and we have to go to secp384r1, or secp521r1?  or any of the sect series?

File: crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.tuweni.units.bigints.UInt256;
+
+import java.math.BigInteger;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class Signature {
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java:15]: Probably should rename it something more specific, like `SECPSignature`.

File: crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+    public static final int BYTES_REQUIRED = 65;
```
Comment (shemnon) [inline crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java:17]: Again, we shouldn't lock in the signature length.  It is reasonable to prepare for chains that may require stronger EC cures.

**Review by shemnon — COMMENTED** [review]
Partial review,  I focused on the main classes and skimmed through the unavoidably long list of import renames.

File: crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignatureFactory.java (hunk 4 above)
Comment (lucassaldanha) [reply crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignatureFactory.java:4]: Correct. We are considering this being a config in the genesis file. Mostly because it makes it easier for other people deploying nodes in the network (they don't have to remember the right config, just use the same genesis). Thoughts?

File: crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java (hunk 11 above)
Comment (lucassaldanha) [reply crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java:17]: I guess the impact of supporting stronger keys might add extra work that won't provide much value atm. One of the assumptions for this is that secp256k1 and secp256r1 use keys of the same size. If we make the size dynamic, we might need to handle this across different layers of the system.

I'm not saying that we shouldn't do it, but maybe not in this first iteration.

File: crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignature.java (hunk 2 above)
Comment (lucassaldanha) [reply crypto/src/main/java/org/hyperledger/besu/crypto/EllipticCurveSignature.java:11]: > I feel we should go a bit further than simply pulling out the interface methods. I think we should reconsider all the APIs being pulled through and look to see if better homes exist (calculateECDHKeyAgreement, getHalfCureOrder, enableNarive, etc).

I believe the idea is to make incremental changes. And because this class is used pretty much everywhere, pulling out the interface already introduces a bunch of changes to review. Do you feel like it is better to have all changes done at once?

> I also think we can do better than the EllipticCurveSignature name. Account sigs may not always be ECDSA (DSA and RSA are also listed in FIPS)

I think supporting "non-EC" have bigger implications that are out of the scope of this PR. I guess this is a small step towards something like https://github.com/EntEthAlliance/client-spec/issues/110. (a really small step)

File: crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java (hunk 11 above)
Comment (shemnon) [reply crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java:17]: We can expose the key length as a signature API and the initial batch of factory methods can presume it.  But I'm wary of baking in 65 bytes too deeply into the API.

File: crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java (hunk 10 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/Signature.java]: Good suggestion. I've renamed it.

File: crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java (hunk 8 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java]: Good suggestion. I've renamed it.

File: crypto/src/main/java/org/hyperledger/besu/crypto/PrivateKey.java (hunk 7 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/PrivateKey.java]: Good suggestion. I've renamed it.

File: crypto/src/main/java/org/hyperledger/besu/crypto/KeyPairUtil.java (hunk 5 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/KeyPairUtil.java]: Thank you for the tip. I've used now `Suppliers#memoize` instead.

File: crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java (hunk 9 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/PublicKey.java]: I would suggest to do that in a seperate PR. I've looked a bit into it and there are some calculation, e.g. to recover the public key, which are affected by that too. I think a seperate (and much smaller) PR is much easier to review such a bit more difficult change.
I will do that after this one is merged.

File: crypto/src/main/java/org/hyperledger/besu/crypto/KeyPair.java (hunk 3 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/KeyPair.java:29]: Looking through the code that would require a lot of casting from `javax.security.PublicKey` -> `SECPPublicKey`, because `SECPPublicKey.getEncodedBytes` is used in a lot of places and not part of the interface. Considering this I would leave the class for right now, as it is.

**Review by lucassaldanha — APPROVED** [review]
I think we can merge this as a first step. This PR is only getting rid of the static references. However, there were some good suggestions that I'd like to see done as well. Let's make sure we work on those next.

**Comment by matkt** [comment]
I have quite a bit of error message while synchronizing my nodes. I got it on Goerli, Ropsten, Mordor.

The error does not seem to prevent the synchronization from doing well but my logs are full of this messsage.

Looking at the stacktrace this seems related to this modification. Any idea ?

> {"timestamp":"2021-04-13T09:35:38,122","level":"ERROR","thread":"vert.x-eventloop-thread-2","class":"VertxPeerDiscoveryAgent","message":"Encountered error while handling packet","throwable":" java.lang.IllegalArgumentException: Could not recover public key\n\tat org.hyperledger.besu.crypto.SECP256K1.recoverFromSignatureNative(SECP256K1.java:495)\n\tat org.hyperledger.besu.crypto.SECP256K1.recoverPublicKeyFromSignature(SECP256K1.java:370)\n\tat org.hyperledger.besu.ethereum.p2p.discovery.internal.Packet.<init>(Packet.java:90)\n\tat org.hyperledger.besu.ethereum.p2p.discovery.internal.Packet.decode(Packet.java:126)\n\tat org.hyperledger.besu.ethereum.p2p.discovery.VertxPeerDiscoveryAgent.lambda$handlePacket$6(VertxPeerDiscoveryAgent.java:221)\n\tat io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)\n\tat io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n"}

**Comment by daniel-iobuilders** [comment]
@matkt Thank you for reporting it. Do you use any non-default parameters for starting Besu? Or do you use the default ones, like for exmple:

```
besu --network=ropsten
```

**Comment by matkt** [comment]
> @matkt Thank you for reporting it. Do you use any non-default parameters for starting Besu? Or do you use the default ones, like for example:
> 
> ```
> besu --network=ropsten
> ```

I use the default parameters. When I run Besu the error does not appear directly we have to wait for the synchronization to progress and then we have the error messages. If I restart Besu it stop and after some time it reappears

**Comment by daniel-iobuilders** [comment]
@matkt Thank you! I could reproduce the error. I look into it and will let you know when I have found the solution.

**Comment by 3nprob** [comment]
@daniel-iobuilders Thanks for looking into it. We're getting the same on fast sync nodes on rinkeby/goerli/mainnet, let me know if there's any particular part of config that's relevant to share

**Comment by daniel-iobuilders** [comment]
@matkt @3nprob The problem is caused by apparently invalid signatures of some packets that the discovery agent receives. I captured some of these packets and created a test to reproduce the error and tried it with different releases of Besu. This showed that this error is not related to this PR, but also occurs in Besu versions that were released before it, like 21.1.0 or 20.10.4. This leads to the conclusion that the packets are indeed invalid and they don't affect the correct functionality of Besu.

You can verify this by simply running one of the mentioned releases. After a few minutes you will see the same error messages in the logs.

To summarize: This error message appears when a packet of the discovery protocol has an invalid signature. Besu tries to verify it, which throws the exception in question. This is not really an error, because in an open network you can always receive invalid packets and it does not affect any functionality. So what needs to be changed is, how Besu handles those packets. I would suggest that instead of throwing an error a simply debug message in the logs is sufficient. Like this the logs don't get cluttered by it. I will open an issue, share my findings and propose this solution to avoid those messages in the future.
